python/amici/pandas.py

```diff
@@ -189,226 +262,272 @@ def _fill_conditions_dict(datadict, model, edata) -> dict:
-        _get_names_or_ids(model, 'Observable')
+         _get_names_or_ids(model, 'FixedParameter', by_id=by_id)] + \
+        _get_names_or_ids(model, 'Observable', by_id=by_id)
 
 
-def _get_state_cols(model):
-    """ Construction helper for state dataframe headers
+def _get_state_cols(model, by_id):
+    """ Construction helper for state dataframe headers.
 
     Arguments:
-        model: Model instance
+        model: Model instance.
+        by_id: bool
+            If True, ids are used as identifiers, otherwise the possibly more
+            descriptive names.
 
     Returns:
-        column names as list
+        column names as list.
 
     Raises:
 
     """
     return \
         ['time', 't_presim'] + \
-        _get_names_or_ids(model, 'FixedParameter') + \
+        _get_names_or_ids(model, 'FixedParameter', by_id=by_id) + \
         [name + '_preeq' for name in
-         _get_names_or_ids(model, 'FixedParameter')] + \
+            _get_names_or_ids(model, 'FixedParameter', by_id=by_id)] + \
         [name + '_presim' for name in
-         _get_names_or_ids(model, 'FixedParameter')] + \
-        _get_names_or_ids(model, 'State')
+            _get_names_or_ids(model, 'FixedParameter', by_id=by_id)] + \
+        _get_names_or_ids(model, 'State', by_id=by_id)
 
 
-def _get_names_or_ids(model, variable):
-    """ Obtains a unique list of identifiers for the specified variable
-        first tries model.getVariableNames and then uses model.getVariableIds
+def _get_names_or_ids(model, variable, by_id):
+    """
+    Obtains a unique list of identifiers for the specified variable.
+    First tries model.getVariableNames and then uses model.getVariableIds.
 
     Arguments:
-        model: Model instance
-        variable: variable name
+        model: Model instance.
+        variable: variable name.
+        by_id: bool
+            If True, ids are used as identifiers, otherwise first the possibly
+            more descriptive names are used.
 
     Returns:
-        column names as list
+        column names as list.
 
     Raises:
 
     """
+    # check whether variable type permitted
     variable_options = ['Parameter', 'FixedParameter', 'Observable', 'State']
     if variable not in variable_options:
         raise ValueError('variable must be in ' + str(variable_options))
+
+    # functions to extract attributes
     namegetter = getattr(model, 'get' + variable + 'Names')
     idgetter = getattr(model, 'get' + variable + 'Ids')
-    if len(set(namegetter())) == len(namegetter()) \
+
+    # extract labels
+    if not by_id and len(set(namegetter())) == len(namegetter()) \
             and model.hasObservableNames():
+        # use variable names
         return list(namegetter())
     elif model.hasObservableIds():
+        # use variable ids
         return list(idgetter())
     else:
+        # unable to create unique lables
         raise RuntimeError('Model Observable Names are not unique and '
```

Comment:
I think we should customize the Exception text according to the value of `by_id`

Reply:
Also just noticed that the function was generalized to different variable without updating the exception text, replace 
```suggestion
        raise RuntimeError(f'Model {variable} names are not unique and '
```

Reply:
in the method, also ``model.hasObservableIds()`` is called. That would also not fit for general variables, right?

Reply:
Yes! Needs to be replaced by `has{Variable}Names()`. See http://icb-dcm.github.io/AMICI/classamici_1_1_model.html